Add boilerplate draft for Central API Tech Preview blog post (#990)
SovereignCloudStack/website: _i18n/en/_posts/blog/2024-08-13-central-api-tech-preview-release.md

    subgraph "With central API (simplified)"
        User2{"User"}
        subgraph "provider responsibility"
            CentralAPI["Central API"]
            OpenStack2["OpenStack API"]
            Keycloak2["Keycloak API"]
            CAPI2["Cluster API"]
        end

        User2
            -- uses --> K8sTooling2["kubectl/\nargocd/flux/..."]
        K8sTooling2 -- calls --> CentralAPI
        CentralAPI -- calls --> OpenStack2
        CentralAPI -- calls --> Keycloak2
        CentralAPI -- calls --> CAPI2
    end
    subgraph "Without central API (simplified)"
        User1{"User"}
        subgraph "provider responsibility"
            OpenStack1["OpenStack API"]
            Keycloak1["Keycloak API"]
        end
        CAPI1["Cluster API"]

        User1
            -- uses --> OpenStackCLI1["OpenStackCLI/OpenStackUI/\nTerraform/Ansible/..."]
            -- calls --> OpenStack1
        User1
            -- uses --> KeycloakCLI1["KeycloakCLI/KeycloakUI/\nTerraform/Ansible/..."]
            -- calls --> Keycloak1
        User1
            -- uses --> K8sTooling1["kubectl/\nargocd/flux/..."]
            -- calls --> CAPI1
    end
```

## What is the SCS Central API

> We wanted to build a common interface for cloud endusers to manage their infrastructure centrally with one API. This is a huge task when compared to already existing solutions from other clouds. As we face multiple complex APIs we need to unify, like OpenStack and KeyCloak, abstracting everything would be a challenging task. We identified several problems we need to solve which a basically: Authentication, Authorization, Multi-Tenancy, and the notion of resources. The first two, we already mentioned before. Multi-Tenancy is of utmost importance as a single Central API Instance should be available to all users of a cloud and a tenant should not interfer with resources of another tenant in any way. The notion of resources is the last important issue as the user experience should be uniform and requests to the Central API should feel the same regardless of deploying a VM or creating a Kuberetes Cluster. We discussed several approaches on how to build such an interface.

Reviewer: ```suggestion
We wanted to build a common interface for cloud end users to manage their infrastructure centrally with one API. This is a huge task when compared to already existing solutions from other clouds. As we face multiple complex APIs we need to unify, like OpenStack and Keycloak, abstracting everything would be a challenging task. We identified several problems we need to solve which a basically: Authentication, Authorization, Multi-Tenancy, and the notion of resources. The first two, we already mentioned before. Multi-Tenancy is of utmost importance as a single Central API Instance should be available to all users of a cloud and a tenant should not interfere with resources of another tenant in any way. The notion of resources is the last important issue as the user experience should be uniform and requests to the Central API should feel the same regardless of deploying a VM or creating a Kubernetes Cluster. We discussed several approaches on how to build such an interface.
```

1. Building a Restful API with a Web Framework
2. Extending the OpenStack API with a custom OpenStack Driver
3. Building ontop of the Kubernetes API

> Building a Central API as a new Restful API would offer most freedom on designing the API. The problem would be that we would reinventing some wheels along the way as all expects need to be created. In the SCS Project, we strive to use upstream components are strengthen upstream projects by not only using their work, but to contribute and upstream our work to make it available for others to use. Creating a new service would contradict this approach and would lead to a huge effort creating the resource definitions, implementing Authorization and Authorization. Building an own Multi-Tenancy aproach would also be complicated as this would be security relevant and needed to be audited.

Reviewer: ```suggestion
Building a Central API as a new Restful API would offer most freedom on designing the API. The problem would be that we would reinventing some wheels along the way as all aspects need to be created. In the SCS Project, we strive to use upstream components and strengthen them by not only using their work, but to contribute and upstream our work to make it available for others to use. Creating a new service would contradict this approach and would lead to a huge effort creating the resource definitions, implementing Authorization and Authorization. Building an own Multi-Tenancy approach would also be complicated as this would be security relevant and needed to be audited.
```

Using the OpenStack API as the chosen standard would be a viable approach to manage Kubernetes Clusters and other resources as projects like Magnum show. The SCS Project's [technological vision](https://scs.community/about/#technological-vision) does not see OpenStack as the primary service and we want to avoid vendor lock-ins as far as possible. So this approach was also dismissed.

> The third approach is using the Kubernetes API as the Central API. This would solve some issues right from the start as already existing concepts of Kubernets could be used off the shelve like Authentication and Authorization with Kubernetes RBAC. Using Kubernetes also gives us huge and generic ecosystem. The Kubernetes API comes with an built-in notion of resources. Resources have a Kind, Metadata and a Spec and can be defined as Custom Resources. Multi-Tenancy can be achived by using Namespaces. Together with using the Kubernetes API we choose to also leverage [Crossplane](https://www.crossplane.io/), which is a Platform Engineering tool. It is used to provision cloud resources by using Kubernetes nativ resources. The Platform Team creates so called Crossplane Resource Definitions (XRDs) which creates Compositions and Claims. Compositions are templated mananged resources, which are in turn Kubernetes resources descibing real cloud resources outside of the Kubernetes Cluster. Those resources are cluster scoped. Claims on the other hand are namespaced and are used by a team, or in our case a tenant, to manage their cloud resources. This way, tenant resources are isolated and multi-tenancy is achieved.

Reviewer: ```suggestion
The third approach is using the Kubernetes API as the Central API. This would solve some issues right from the start as already existing concepts of Kubernetes could be used off the shelf, like Authentication and Authorization with Kubernetes RBAC. Using Kubernetes also gives us a huge and generic ecosystem. The Kubernetes API comes with an built-in notion of resources. Resources have a Kind, Metadata and a Spec and can be defined as Custom Resources. Multi-Tenancy can be achived by using Namespaces. Together with using the Kubernetes API we choose to also leverage [Crossplane](https://www.crossplane.io/), which is a Platform Engineering tool. It is used to provision cloud resources by using Kubernetes native resources. The Platform Team creates so called Crossplane Resource Definitions (XRDs) which creates Compositions and Claims. Compositions are templated mananged resources, which are in turn Kubernetes resources describing real cloud resources outside of the Kubernetes Cluster. Those resources are cluster scoped. Claims on the other hand are namespaced and are used by a team, or in our case a tenant, to manage their cloud resources. This way, tenant resources are isolated and multi-tenancy is achieved.
```

The following tables summerizes the different aproaches and efforts.

| Resource Models/Patterns | API Backend | AuthN | AuthZ | Multi-tenancy |
| -------- | -------- | --- | --- | ---- |
| Selfdefined API's | Selfmade REST server | Selfmade | Selfmade | Selfmade |
| OpenStack API's | Selfmade REST server? | Keystone Integration? | Keystone Integration? | Keystone Integration? |
| Kubernetes/Crossplane API | Kubernetes apiserver | Kubernetes apiserver AuthN | Kubernetes RBAC | Crossplane namespacing |

| Resource Models/Patterns | Amount of base work | Amount of work per resource | Ease of adoption / ecosystem | Scalability |
| -------- | -------- | --- | --- | --- |

> | Selfdefined API's | <span style="color: orange">Unknown</span>, depends on tech stack | <span style="color: orange">Unknown</span> (Maybe <span style="color: green">Low</span> if utilizing `fog`/...) | <span style="color: red">No ecosystem</span>| <span style="color: green">Good</span>, depending on architecture |

Reviewer: The coloring (which I used for the Hedgedoc presentation mode) somehow is not rendered in the Github MD previews. Maybe it will appear on the actual website, again?

Author: I don't know, depends on the Markdown renderer. Just try it out, it seems to be removed in either case. (Github MD renderer knows it's HTML and removes the tags, but doesn't render them)

| OpenStack API's | <span style="color: orange">Unknown</span> | <span style="color: orange">Unknown</span> (Maybe <span style="color: green">Low</span> if utilizing `fog`/...) | <span style="color: green">Good</span> for existing API's, <span style="color: red">bad</span> for others | <span style="color: green">Good</span>, probably |
| Kubernetes/Crossplane API | <span style="color: green">Low</span>, utilizing existing concepts | <span style="color: green">Low</span> due to existing Crossplane providers | <span style="color: green">Great</span>, due to huge, generic ecosystem for any resource | <span style="color: orange">Alright</span> (scale with namespaces or clusters) |

## What is the current status of the Central API

> As of now, we build a first technical preview that is capable of managing OpenStack Resources as well as creating Kubernetes Clusters with ClusterStacks as well as Gardener. For OpenStack resources we used the upstream OpenStack Crossplane Provider as it gives us control over a numerous amount of OpenStack resources. These are not feature complete but resources a user commonly would use are present and usable. A Cloud Service Provider need to set up the the OpenStack credentials for a tenant namespace. The SCS XRD are mirrored from the OpenStack Provider and are already bundled as a Crossplane package to be used.

Reviewer: ```suggestion
As of now, we build a first technical preview that is capable of managing OpenStack Resources as well as creating Kubernetes Clusters with ClusterStacks as well as Gardener. For OpenStack resources we used the upstream OpenStack Crossplane Provider as it gives us control over a numerous amount of OpenStack resources. These are not feature complete but resources a user commonly would use are present and usable. A Cloud Service Provider needs to set up the the OpenStack credentials for a tenant namespace. The SCS XRD's are mirrored from the OpenStack Provider and are already bundled as a Crossplane package to be used.
```

> For managing Kubernetes Clusters, we used the Crossplane Kubernetes Provider to create Manifests on remote Kubernetes Clusters. The XRDs created will create the Compositions and Claims to create ClusterStack Manifests on a ClusterAPI Management Cluster. Hence, a user will be able to use such a Claim to trigger the creation and deletion of such ClusterAPI Clusters. Additionally, we created such an XRD to also support Gardener Clusters by creating Gardener Shoot Manifests on Gardener Clusters. Currently, it is only possible to set the Kubernetes Version in the Claim Resource, but it proves that it is possible to have manage all those Cloud resources over a single API that is open source and leverages upstream work.

Reviewer: ```suggestion
For managing Kubernetes Clusters, we created a very simple generic "Cluster" XRD.
When combined with a corresponding Gardener composition, shoot resources can be applied to a Gardener Kubernetes API - using the Crossplane Kubernetes Provider. The same works for creating ClusterAPI Clusters on a Cluster API Management Cluster. Hence, a user will be able to use such a Claim to trigger the creation and deletion of such ClusterAPI Clusters. Currently, it is only possible to set the Kubernetes Version in the Claim Resource, but it proves that it is possible to have manage all those Cloud resources over a single API that is open source and leverages upstream work.
```
